File: vector/zig/vector.zig
const std = @import("std");

const Random = struct {
    const IM: usize = 139968;
    const IMF: f64 = 139968.0;
    const IA: usize = 3877;
    const IC: usize = 29573;
    seed: usize,

    pub fn init() Random {
        return .{ .seed = 42 };
    }

    pub fn next(self: *Random, max: usize) usize {
        self.seed = (self.seed * IA + IC) % IM;
        return @intFromFloat(@as(f64, @floatFromInt(max)) * (@as(f64, @floatFromInt(self.seed)) / IMF));
    }
};

pub fn main(init: std.process.Init) !void {
    const allocator = init.gpa;
    var numbers = try std.ArrayList(usize).initCapacity(allocator, 16);
    defer numbers.deinit(allocator);

    var rand = Random.init();
    const max: usize = 100;
    const size: usize = 200000000;

    var i: usize = 0;
    while (i < size) : (i += 1) {
        try numbers.append(allocator, rand.next(max));
    }

    var sum: usize = 0;
    i = 0;
    while (i < 1000) : (i += 1) {
        sum += numbers.items[rand.next(size)];
    }

    var stdout_writer = std.Io.File.stdout().writer(init.io, &.{});
    const stdout = &stdout_writer.interface;
    try stdout.print("{}\n", .{sum});
}

File: vector/ocaml/vector.ml
type random = { im : int; imf : float; ia : int; ic : int; mutable seed : int }

let create_random () =
  { im = 139968; imf = 139968.0; ia = 3877; ic = 29573; seed = 42 }

let next rand max_val =
  rand.seed <- ((rand.seed * rand.ia) + rand.ic) mod rand.im;
  let max_f = float_of_int max_val in
  let seed_f = float_of_int rand.seed in
  int_of_float (max_f *. (seed_f /. rand.imf))

let () =
  let rand = create_random () in
  let max_val = 100 in
  let size = 200_000_000 in

  let numbers = Dynarray.create () in
  for _ = 1 to size do
    Dynarray.add_last numbers (next rand max_val)
  done;

  let sum = ref 0 in
  for _ = 1 to 1000 do
    let idx = next rand size in
    sum := !sum + Dynarray.get numbers idx
  done;

  Printf.printf "%d\n" !sum
